src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -193,39 +200,59 @@ public override void OnWriterCompleted(Action<Exception, object> callback, objec
         }
 
         /// <inheritdoc />
-        public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
+        public override ValueTask<ReadResult> ReadAsync(CancellationToken cancellationToken = default)
         {
             // TODO ReadyAsync needs to throw if there are overlapping reads.
             ThrowIfCompleted();
 
-            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
-            CancellationTokenSource tokenSource = InternalTokenSource;
-            if (TryReadInternal(tokenSource, out ReadResult readResult))
+            if (TryReadInternal(out ReadResult readResult))
             {
-                return readResult;
+                return new ValueTask<ReadResult>(readResult);
             }
 
             if (_isStreamCompleted)
             {
-                return new ReadResult(buffer: default, isCanceled: false, isCompleted: true);
+                return new ValueTask<ReadResult>(new ReadResult(buffer: default, isCanceled: false, isCompleted: true));
+            }
+
+            if (_streamReadTask.IsCompleted)
+            {
+                _streamReadTask = ReadStreamAsync(cancellationToken);
+
+                // Completed the stream read inline because it was synchronous and there was no exception thrown
+                if (_streamReadTask.IsCompleted && _edi == null)
+                {
+                    return new ValueTask<ReadResult>(GetReadResult());
+                }
             }
 
-            var reg = new CancellationTokenRegistration();
+            return new ValueTask<ReadResult>(this, 0);
+        }
+
+        private async Task ReadStreamAsync(CancellationToken cancellationToken)
+        {
+            CancellationTokenSource cts = null;
+            CancellationToken effectiveToken;
+
             if (cancellationToken.CanBeCanceled)
             {
-                reg = cancellationToken.UnsafeRegister(state => ((StreamPipeReader)state).Cancel(), this);
+                cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _completeCts.Token);
```

Comment:
Why do we need to create a LinkedTokenSource instead of doing the cancel pattern we had before?